File: backend/models/immunization.go
package models

import (
	"time"
)

// ImmunizationSchedule represents a master immunization schedule item
type ImmunizationSchedule struct {
	ID                      string    `json:"id" db:"id"`
	Name                    string    `json:"name" db:"name"`
	NameID                  *string   `json:"name_id,omitempty" db:"name_id"`
	Description             *string   `json:"description,omitempty" db:"description"`
	
	// Timing in days
	AgeMinDays              *int      `json:"age_min_days,omitempty" db:"age_min_days"`
	AgeOptimalDays          *int      `json:"age_optimal_days,omitempty" db:"age_optimal_days"`
	AgeMaxDays              *int      `json:"age_max_days,omitempty" db:"age_max_days"`
	
	// Timing in months (for display)
	AgeMinMonths            *int      `json:"age_min_months,omitempty" db:"age_min_months"`
	AgeOptimalMonths        *int      `json:"age_optimal_months,omitempty" db:"age_optimal_months"`
	AgeMaxMonths            *int      `json:"age_max_months,omitempty" db:"age_max_months"`
	
	// Dose information
	DoseNumber              int       `json:"dose_number" db:"dose_number"`
	TotalDoses              *int      `json:"total_doses,omitempty" db:"total_doses"`
	
	// Interval information
	IntervalFromPreviousDays *int     `json:"interval_from_previous_days,omitempty" db:"interval_from_previous_days"`
	IntervalFromPreviousMonths *int   `json:"interval_from_previous_months,omitempty" db:"interval_from_previous_months"`
	
	// Category & priority
	Category                string    `json:"category" db:"category"` // wajib, tambahan
	Priority                string    `json:"priority" db:"priority"` // high, medium, low
	IsRequired              bool      `json:"is_required" db:"is_required"`
	
	// Additional info
	Notes                   *string   `json:"notes,omitempty" db:"notes"`
	Source                  string    `json:"source" db:"source"`
	IsActive                bool      `json:"is_active" db:"is_active"`
	
	CreatedAt               time.Time `json:"created_at" db:"created_at"`
	UpdatedAt               time.Time `json:"updated_at" db:"updated_at"`
}

// ChildImmunization represents a recorded immunization for a child
type ChildImmunization struct {
	ID                      string    `json:"id" db:"id"`
	ChildID                 string    `json:"child_id" db:"child_id"`
	ImmunizationScheduleID  string    `json:"immunization_schedule_id" db:"immunization_schedule_id"`
	
	// Given date and age
	GivenDate               string    `json:"given_date" db:"given_date"` // YYYY-MM-DD
	GivenAtAgeDays          *int      `json:"given_at_age_days,omitempty" db:"given_at_age_days"`
	GivenAtAgeMonths        *int      `json:"given_at_age_months,omitempty" db:"given_at_age_months"`
	
	// Details
	Location                *string   `json:"location,omitempty" db:"location"`
	HealthcareFacility      *string   `json:"healthcare_facility,omitempty" db:"healthcare_facility"`
	DoctorName              *string   `json:"doctor_name,omitempty" db:"doctor_name"`
	VaccineBatchNumber      *string   `json:"vaccine_batch_number,omitempty" db:"vaccine_batch_number"`
	Notes                   *string   `json:"notes,omitempty" db:"notes"`
	
	// Status
	IsOnSchedule            *bool     `json:"is_on_schedule,omitempty" db:"is_on_schedule"`
	IsCatchUp               bool      `json:"is_catch_up" db:"is_catch_up"`
	
	CreatedAt               time.Time `json:"created_at" db:"created_at"`
	UpdatedAt               time.Time `json:"updated_at" db:"updated_at"`
	
	// Joined fields
	Schedule                *ImmunizationSchedule `json:"schedule,omitempty" db:"-"`
}

// ImmunizationRecordRequest is the payload for recording an immunization
type ImmunizationRecordRequest struct {
	ImmunizationScheduleID string  `json:"immunization_schedule_id" validate:"required"`
	GivenDate              string  `json:"given_date" validate:"required"` // YYYY-MM-DD
	Location               *string `json:"location,omitempty"`
	HealthcareFacility     *string `json:"healthcare_facility,omitempty"`
	DoctorName             *string `json:"doctor_name,omitempty"`
	VaccineBatchNumber     *string `json:"vaccine_batch_number,omitempty"`
	Notes                  *string `json:"notes,omitempty"`
}

// ImmunizationStatus represents the status of an immunization for a child
type ImmunizationStatus struct {
	Schedule                ImmunizationSchedule `json:"schedule"`
	Status                  string               `json:"status"` // pending, completed, overdue, upcoming
	DueDate                 *string              `json:"due_date,omitempty"` // YYYY-MM-DD
	DueAgeMonths            *int                 `json:"due_age_months,omitempty"`
	DaysUntilDue            *int                 `json:"days_until_due,omitempty"`
	DaysOverdue             *int                 `json:"days_overdue,omitempty"`
	
	// If completed
	Record                  *ChildImmunization   `json:"record,omitempty"`
}

// ImmunizationScheduleResponse represents the API response for immunization schedule
type ImmunizationScheduleResponse struct {
	ChildID       string               `json:"child_id"`
	AgeMonths     int                  `json:"age_months"`
	AgeDays       int                  `json:"age_days"`
	Immunizations []ImmunizationStatus `json:"immunizations"`
	Summary       ImmunizationSummary  `json:"summary"`
}

// ImmunizationSummary contains summary statistics
type ImmunizationSummary struct {
	Total     int `json:"total"`
	Completed int `json:"completed"`
	Pending   int `json:"pending"`
	Overdue   int `json:"overdue"`
	Upcoming  int `json:"upcoming"`
}

File: frontend/stores/immunization.ts
import { defineStore } from 'pinia'

export interface ImmunizationSchedule {
  id: string
  name: string
  name_id?: string
  description?: string
  age_min_days?: number
  age_optimal_days?: number
  age_max_days?: number
  age_min_months?: number
  age_optimal_months?: number
  age_max_months?: number
  dose_number: number
  total_doses?: number
  interval_from_previous_days?: number
  category: string
  priority: string
  is_required: boolean
  notes?: string
  source: string
}

export interface ChildImmunization {
  id: string
  child_id: string
  immunization_schedule_id: string
  given_date: string
  given_at_age_days?: number
  given_at_age_months?: number
  location?: string
  healthcare_facility?: string
  doctor_name?: string
  vaccine_batch_number?: string
  notes?: string
  is_on_schedule?: boolean
  is_catch_up: boolean
  schedule?: ImmunizationSchedule
}

export interface ImmunizationStatus {
  schedule: ImmunizationSchedule
  status: 'pending' | 'completed' | 'overdue' | 'upcoming'
  due_date?: string
  due_age_months?: number
  days_until_due?: number
  days_overdue?: number
  record?: ChildImmunization
}

export interface ImmunizationSummary {
  total: number
  completed: number
  pending: number
  overdue: number
  upcoming: number
}

export interface ImmunizationScheduleResponse {
  child_id: string
  age_months: number
  age_days: number
  immunizations: ImmunizationStatus[]
  summary: ImmunizationSummary
}

export const useImmunizationStore = defineStore('immunization', {
  state: () => ({
    schedule: [] as ImmunizationStatus[],
    summary: null as ImmunizationSummary | null,
    ageMonths: 0,
    ageDays: 0,
    loading: false,
    error: null as string | null,
  }),
  
  getters: {
    hasSchedule: (state) => state.schedule.length > 0,
    
    completedImmunizations: (state) => 
      state.schedule.filter(i => i.status === 'completed'),
    
    pendingImmunizations: (state) => 
      state.schedule.filter(i => i.status === 'pending'),
    
    overdueImmunizations: (state) => 
      state.schedule.filter(i => i.status === 'overdue'),
    
    upcomingImmunizations: (state) => 
      state.schedule.filter(i => i.status === 'upcoming'),
    
    immunizationsByCategory: (state) => {
      const grouped: Record<string, ImmunizationStatus[]> = {}
      state.schedule.forEach(imm => {
        const cat = imm.schedule.category || 'lainnya'
        if (!grouped[cat]) {
          grouped[cat] = []
        }
        grouped[cat].push(imm)
      })
      return grouped
    },
    
    immunizationsByPriority: (state) => {
      const grouped: Record<string, ImmunizationStatus[]> = {
        high: [],
        medium: [],
        low: []
      }
      state.schedule.forEach(imm => {
        const priority = imm.schedule.priority || 'medium'
        if (grouped[priority]) {
          grouped[priority].push(imm)
        }
      })
      return grouped
    },
  },
  
  actions: {
    async fetchSchedule(childId: string) {
      if (!childId) {
        this.schedule = []
        this.summary = null
        return
      }
      
      this.loading = true
      this.error = null
      
      try {
        const authStore = useAuthStore()
        const apiBase = useApiUrl()
        
        const data = await $fetch<ImmunizationScheduleResponse>(
          `${apiBase}/api/children/${childId}/immunizations`,
          {
            headers: {
              'Authorization': `Bearer ${authStore.token}`
            }
          }
        )
        
        this.schedule = data.immunizations || []
        this.summary = data.summary || null
        this.ageMonths = data.age_months || 0
        this.ageDays = data.age_days || 0
      } catch (e: any) {
        console.error('Failed to fetch immunization schedule:', e)
        this.error = e.data?.error || e.message || 'Failed to fetch immunization schedule'
        this.schedule = []
        this.summary = null
      } finally {
        this.loading = false
      }
    },
    
    async recordImmunization(childId: string, record: {
      immunization_schedule_id: string
      given_date: string
      location?: string
      healthcare_facility?: string
      doctor_name?: string
      vaccine_batch_number?: string
      notes?: string
    }) {
      this.loading = true
      this.error = null
      
      try {
        const authStore = useAuthStore()
        const apiBase = useApiUrl()
        
        await $fetch(
          `${apiBase}/api/children/${childId}/immunizations`,
          {
            method: 'POST',
            headers: {
              'Authorization': `Bearer ${authStore.token}`,
              'Content-Type': 'application/json'
            },
            body: record
          }
        )
        
        // Refresh schedule after recording
        await this.fetchSchedule(childId)
        
        return { success: true }
      } catch (e: any) {
        console.error('Failed to record immunization:', e)
        this.error = e.data?.error || e.message || 'Failed to record immunization'
        return { success: false, error: this.error }
      } finally {
        this.loading = false
      }
    },
    
    clearSchedule() {
      this.schedule = []
      this.summary = null
      this.ageMonths = 0
      this.ageDays = 0
      this.error = null
    },
  },
})

